scripts/util/doc-builder.js

```diff
@@ -0,0 +1,341 @@
+  NodeGit.Repository.open('.').then(function (repo) {
+
+    self.log('info', 'Getting available references within the repo');
+    return repo.getReferences(NodeGit.Reference.TYPE.OID).then(function(references){
+      var treesDataPromises = [];
+
+      self.log('info', 'Filtering them to include only tags and the ', self.options.masterBranchName, ' branch.');
+      references.forEach(function(reference){
+        var target = reference.targetPeel() || reference.target();
+        var name = reference.shorthand();
+        var commitPromise;
+
+        if(reference.isTag()) {
+          commitPromise = repo.getCommit(target);
+        } else if (reference.isBranch() && reference.shorthand() === self.options.masterBranchName) {
+          commitPromise = repo.getBranchCommit(name);
+        } else {
+          return;
+        }
+        self.log('info', 'Gotten a promise for commit for ', name);
+        self.log('info', 'Adding a promise to get commit tree for ', name);
+        treesDataPromises.push(buildTreePromise(commitPromise, name));
+      });
+
+      self.log('info', 'Resolving commit tree promises');
+      return Promise.all(treesDataPromises);
+    }).then(function(treesData){
+      var paths = [];
+      var options = JSON.parse(JSON.stringify(self.options));
+      options.repo = repo;
+
+      self.log('info', 'Checking out commit trees');
+      checkoutTrees(treesData.shift(), treesData, options, function (paths) {
+        self.log('info', 'Done checking out source code');
+
+        if (callback) {
+          callback(null, paths);
+        }
+      });
+    });
+  }).catch(function(error){
+    console.error("Error building doc");
+    console.error(error);
+    if (callback) {
+      callback(error);
+    }
+  });
+}
+
+DocBuilder.prototype.generateAPIDocs = function (paths, callback) {
+  var self = this;
+
+  self.log('info', '\nGenerating API docs');
+  yuidoc.generate({paths: paths, themeDir: self.options.themeDir}, function () {
+    self.log('info', 'Done generating API docs');
+    if (callback) {
+      callback();
+    }
+  });
+}
+
+DocBuilder.prototype.commitAPIDocs = function (callback) {
+  var self = this;
+  var apiTempDir = path.join('.', self.options.versionsDirName);
+
+  self.log('info', '\nCommiting API Docs');
+  if (fs.existsSync(apiTempDir)) {
+    recursiveRmdir(apiTempDir);
+  }
+
+  self.log('info', 'Copying ', self.options.apiDirPath, ' to root directory');
+  recursiveCpdir(self.options.apiDirPath, apiTempDir);
+
+  NodeGit.Repository.open('.').then(function (repo) {
+
+    self.log('info', 'Getting repo\'s current staging area');
+    return repo.index().then(function (index) {
+      self.log('info', 'Getting ', self.options.docsBranchName, '\'s latest commit');
+      return repo.getBranchCommit(self.options.docsBranchName).then(function (docsBranchCommit) {
+        self.log('info', 'Getting the commit\'s tree');
+        return docsBranchCommit.getTree().then(function (tree) {
```

Comment:
@m-ux, i learnt about `Promise.resolve().then()` after reading the article. I had the other approach of creating new promises earlier. So we would still have absolutely sane code converting the functions to return promises. I'd get right on that.

Thanks.